src/interpret.ml

```diff
@@ -282,63 +294,204 @@ let exec_extern_func stack (f : Value.Func.extern_func) =
   | R4 (t1, t2, t3, t4), (v1, v2, v3, v4) ->
     push_val t1 v1 stack |> push_val t2 v2 |> push_val t3 v3 |> push_val t4 v4
 
-let rec exec_instr env locals stack instr =
+module State = struct
+  type stack = Env.t' Stack.t
+
+  type value = Env.t' Value.t
+
+  type locals = value array
+
+  type pc = (int, func_type) instr' list
+
+  type block =
+    { branch : pc
+    ; branch_rt : result_type
+    ; continue : pc
+    ; continue_rt : result_type
+    ; stack : stack
+    ; is_loop : bool
+    }
+
+  type block_stack = block list
+
+  type exec_state =
+    { return_state : exec_state option
+    ; stack : stack
+    ; locals : locals
+    ; pc : pc
+    ; block_stack : block_stack
+    ; func_rt : result_type
+    ; env : Env.t
+    }
+
+  exception Result of value list
+
+  let return (state : exec_state) =
+    let args = Stack.keep state.stack (List.length state.func_rt) in
+    match state.return_state with
+    | None -> raise (Result args)
+    | Some state ->
+      let stack = args @ state.stack in
+      { state with stack }
+
+  let branch (state : exec_state) n =
+    let rec drop_n l n =
```

Comment:
Can you put this in the `Stack` module ?

Reply:
Hum, actually you added `let rec drops s n =` in `Stack`, so you can use it directly (but I believe `drop_n` to be a better name)